src/Utilities/functions.jl

```diff
@@ -135,43 +143,26 @@ If `x` (resp. `y`, `z`) is `VariableIndex(1)` (resp. 2, 3).
 The canonical representation of `ScalarAffineFunction([y, x, z, x, z], [2, 1, 3, -2, -3], 5)` is `ScalarAffineFunction([x, y], [-1, 2], 5)`.
 
 """
-function canonical(f::SAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> t.variable_index.value)
-    terms = MOI.ScalarAffineTerm{T}[]
-    for t in sorted_terms
-        if !isempty(terms) && t.variable_index == last(terms).variable_index
-            terms[end] = MOI.ScalarAffineTerm(terms[end].coefficient + t.coefficient, t.variable_index)
-        elseif !iszero(t.coefficient)
-            if !isempty(terms) && iszero(last(terms).coefficient)
-                terms[end] = t
-            else
-                push!(terms, t)
-            end
-        end
-    end
-    if !isempty(terms) && iszero(last(terms).coefficient)
-        pop!(terms)
-    end
-    SAF{T}(terms, f.constant)
-end
-function canonical(f::VAF{T}) where T
-    sorted_terms = sort(f.terms, by = t -> (t.output_index, t.scalar_term.variable_index.value))
-    terms = MOI.VectorAffineTerm{T}[]
+function canonical(f::Union{SAF, VAF})
+    sorted_terms = sort(f.terms, by = termindices)
+    terms = eltype(f.terms)[]
     for t in sorted_terms
-        if !isempty(terms) && t.output_index == last(terms).output_index && t.scalar_term.variable_index == last(terms).scalar_term.variable_index
-            terms[end] = MOI.VectorAffineTerm(t.output_index, MOI.ScalarAffineTerm(terms[end].scalar_term.coefficient + t.scalar_term.coefficient, t.scalar_term.variable_index))
-        elseif !iszero(t.scalar_term.coefficient)
-            if !isempty(terms) && iszero(last(terms).scalar_term.coefficient)
+        last_t = last(terms)
+        tnot0 = !iszero(coefficient(t))
+        if !isempty(terms) && termindices(t) == termindices(last_t) && tnot0
```

Comment:
Much nicer :)

Reply:
Although tests are failing somewhere in here

Reply:
Should be fixed now.